src/protocols/rdp/rdp.c

```diff
@@ -227,10 +236,69 @@ static BOOL rdp_freerdp_authenticate(freerdp* instance, char** username,
 
     rdpContext* context = instance->context;
     guac_client* client = ((rdp_freerdp_context*) context)->client;
+    guac_rdp_client* rdp_client = (guac_rdp_client*) client->data;
+    guac_rdp_settings* settings = rdp_client->settings;
+    char* params[4] = {NULL};
+    int i = 0;
+    
+    /* If the client does not support the "required" instruction, warn and
+     * quit.
+     */
+    if (!guac_client_owner_supports_required(client)) {
+        guac_client_log(client, GUAC_LOG_WARNING, "Client does not support the "
+                "\"required\" instruction. No authentication parameters will "
+                "be requested.");
+        return TRUE;
+    }
 
-    /* Warn if connection is likely to fail due to lack of credentials */
-    guac_client_log(client, GUAC_LOG_INFO,
-            "Authentication requested but username or password not given");
+    /* If the username is undefined, add it to the requested parameters. */
+    if (settings->username == NULL) {
+        guac_argv_register(GUAC_RDP_ARGV_USERNAME, guac_rdp_argv_callback, NULL, 0);
+        params[i] = GUAC_RDP_ARGV_USERNAME;
+        i++;
+    }
+    
+    /* If the password is undefined, add it to the requested parameters. */
+    if (settings->password == NULL) {
+        guac_argv_register(GUAC_RDP_ARGV_PASSWORD, guac_rdp_argv_callback, NULL, 0);
+        params[i] = GUAC_RDP_ARGV_PASSWORD;
+        i++;
+    }
+    
+    /* If the domain is undefined, add it to the requested parameters. */
+    if (settings->domain == NULL) {
+        guac_argv_register(GUAC_RDP_ARGV_DOMAIN, guac_rdp_argv_callback, NULL, 0);
+        params[i] = GUAC_RDP_ARGV_DOMAIN;
+        i++;
+    }
```

Comment:
If an administrator provides:

| Parameter | Value |
| --------- | ----- |
| `domain` | (blank) |
| `username` | `foo` |
| `password` | (blank) |

I think the domain is effectively specified - the administrator intends that the user log in to the default domain / local machine, and that the user should be prompted for their password. I doubt that the administrator intends that the user be prompted for their domain and password, with only the username set in stone.

Perhaps we should prompt for the domain _only_ if the username is also blank?

Reply:
Yeah, I'm good with that.